independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java

```diff
@@ -26,6 +27,49 @@
  */
 public final class RunnerClassLoader extends ClassLoader {
 
+    static class BufferView implements ClassLoadingResource.BufferSpan {
+
+        private static final BufferView EMPTY = new BufferView();
+        private byte[] data;
+        private int length;
+
+        // This is used to avoid "moving" an already freshly created buffer when
+        // pushing it back in the stack due to a null JarResource::getResourceData.
+        // if   null    ->  freshly new created one or EMPTY
+        // if   !null   ->  this has been already pushed to the stack: need to move its data
+        private BufferView next;
```

Comment:
N.B. this field is private within the static class, but it's only used by the outer class? Perhaps it's using the wrong visibility, or perhaps you'd expose its state changes via explicit methods instead?

Also - seems like multiple threads might need to read it? I think I understand - roughly speaking - what you have in mind but I feel like it should be spelled out clearly.

Reply:
I can better explain the intent in the code here, let me add more info why is like that; but a preview here...
It's an "intrusive" node ie which contains next pointer into the class itself, to avoid both an indirection and an allocation (and because Java Object Layout reported there was room to fit another field, hence win-win).
The field is "stable" from the pov of who acquire/release (ie pop/push into the concurrent stack of pooled buffers) and it is used for 2 different purposes:
- avoid GC nepotism (unlinking the node from the stack, set it to itself - it could be either be `null` - in next point why it is not)
- it is used to save an additional allocation when a pooled instance is freshly allocated to read a jar and is immediately released back, without filling it with data: the common logic to avoid ABA problems will force a "move" of its byte[] content (ie content is owned by a new freshly allocated node/wrapper), but the `next` pointer can be used to distinguish between nodes which have been already recycled and new ones, saving to create a new node

Reply:
Best have these as comments in the code :)